_posts/2022-1-19-quantization-in-practice.md

```diff
@@ -0,0 +1,199 @@
+
+
+## In PyTorch
+
+PyTorch allows you a few different ways to quantize your model on the CPU, depending on
+- if you prefer a manual, or a more automatic process (*Eager Mode* v/s *FX Graph Mode*)
+- if $S, Z$ for quantizing activations (layer outputs) are precomputed for all inputs, or calculated afresh with each input (*static* v/s *dynamic*),
+- if $S, Z$ are computed during, or after training (*quantization-aware training* v/s *post-training quantization*)
+
+Each approach has its unique tradeoffs; for instance FX Graph Mode can automagically figure out the right quantization configurations, but only for models that are [symbolically traceable](https://PyTorch.org/docs/stable/fx.html#torch.fx.symbolic_trace). Dynamic quantization can offer better precision at the cost of additional overhead in each inference. 
+
+### Post-Training Dynamic Quantization 
+The model's weights are pre-quantized before inference, but the activations are calibrated and quantized on the fly during inference. The simplest of all approaches, it has a one line API call in `torch.quantization.quantize_dynamic`
+
+Because the calibrations are bespoke, clipping ranges can be tighter; dynamic quantization can therefore theoretically give higher accuracies, but calibrating and quantizing each layer's activations can add to the overhead. 
+
+For this reason, it is best suited for models where most of the execution time is spent in loading weights from memory (think very large models with billions of parameters). 
+
+```python
+import torch
+from torch import nn
+
+# toy model
+m = nn.Sequential(
+  nn.Conv1d(2, 64, (8,)),
+  nn.ReLU(),
+  nn.Linear(16,10),
+  nn.LSTM(10, 10))
+
+m.eval()
+
+## EAGER MODE
+from torch.quantization import quantize_dynamic
+model_quantized = quantize_dynamic(
+    model=m, qconfig_spec={nn.LSTM, nn.Linear}, dtype=torch.qint8, inplace=False
+)
+
+## FX MODE
+from torch.quantization import quantize_fx
+qconfig_dict = {"": torch.quantization.default_dynamic_qconfig}  # An empty key denotes the default applied to all modules
+model_prepared = quantize_fx.prepare_fx(m, qconfig_dict)
+model_quantized = quantize_fx.convert_fx(model_prepared)
+```
+
+### Post-Training Static Quantization
+Similar to dynamic quantization, weights are pre-quantized. In dynamic, the activations are calibrated and quantized at inference time. In static, activations are precalibrated by passing sample inputs to the model. 
+
+This method has faster inference than dynamic quantization, but is less robust to accuracy drops from out-of-distribution inputs, i.e. if the model encounters data different from the sample inputs it has calibrated.
```

Comment:
A bit faster or much faster? Also above you recommend dynamic quantization for large models

And a lack of robustness against out of distribution inputs is a big deal, is this technique made for models that need to be fast but hardly change? At least based on what you describe it seems net worse than dynamic quantization

Reply:
the first question is a bit tricky to answer because that can depend on the model architecture. generally, this is recommended for models that have large activations bottlenecking compute and memory access eg: CNNs

yes, it is more susceptible to accuracy issues from drift, possibly ameliorated by regular re-calibration. i'll make this point clearer. thanks!